### PR TITLE
[HOTFIX] WORKAROUND_ISSUE_2038: Disable FP16 and BF16 in ConvHipImplicitGemmV4R1Fwd for MI100/200 and all new targets.

### PR DESCRIPTION
**Pull request by atamazov**
This is expected to unblock CI. Introduces workaround for issue #2038. Related to issue #936.

### ⚠️ May lead to performance regressions!

----
[Attribution] @junliume @johnny-keker
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/urgency_blocker
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/workaround
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/correctness
- Proposed reviewers: 
  - @JehandadKhan
  - @junliume
  - @carlushuang
  - @asroy

### CODE REVIEW

**Review by JehandadKhan — DISMISSED** [review]

**Comment by junliume** [comment]
@atamazov this issue seems to be able to workaround by https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1957/commits/277e35c10f036833f781477f3db7bcb7ad52906d and it is caused by hipRTC header file warning messages are treated as error when `-DBUILD_DEV=On`. Could you help to investigate why it is happening now? Differentiating between passing and failing points have misguided our investigation for a long time.

**Comment by atamazov** [comment]
> @atamazov this issue seems to be able to workaround by [277e35c](https://github.com/ROCmSoftwarePlatform/MIOpen/commit/277e35c10f036833f781477f3db7bcb7ad52906d) and it is caused by hipRTC header file warning messages are treated as error when `-DBUILD_DEV=On`. 

Your observations do not match to what I see the the CI log of #2033:
```
[2023-03-23T18:53:48.490Z]  73/323 Test #128: smoke_solver_ConvHipImplicitGemmV4R1 
...
[2023-03-23T18:53:48.490Z] MIOpen(HIP): Warning [GetAllConfigs] ConvHipImplicitGemmV4R1Fwd: Searching the best solution among 33...
[2023-03-23T18:53:48.490Z] MIOpen(HIP): Warning [Monitor] 0/0/5 0.23334, best within recent 1: 0.23334 #0 16,32,8,2,2,2,4,4,2,4,8,1,16,1,4,32, ETA:0 sec.
[2023-03-23T18:53:48.490Z] MIOpen(HIP): Warning [Monitor] 1/0/5 0.23334, best within recent 1: 0.750709 #1 16,32,4,2,4,4,2,4,2,4,4,1,16,1,4,16, ETA:53.7877 sec.
[2023-03-23T18:53:48.490Z] MIOpen(HIP): Warning [Monitor] 2/0/5 0.23334, best within recent 1: 0.285917 #2 16,128,4,2,4,4,4,4,4,4,4,2,16,2,2,128, ETA:28.3322 sec.
[2023-03-23T18:53:48.490Z] MIOpen(HIP): Warning [Monitor] 3/0/5 0.114943, best within recent 1: 0.114943 #3 16,128,16,2,4,4,4,4,4,4,16,1,16,1,2,128, ETA:15.7516 sec.
[2023-03-23T18:53:48.490Z] MIOpen(HIP): Warning [GenericSearch] Done: 5/0/5, best #3 0.114943 16,128,16,2,4,4,4,4,4,4,16,1,16,1,2,128
[2023-03-23T18:53:48.490Z] MIOpen(HIP): Warning [GenericSearch] ...Score: 2.44989 (default time 0.281597)
[2023-03-23T18:53:48.490Z] FAILED: 0.535381
[2023-03-23T18:53:48.490Z] Max diff: 255
[2023-03-23T18:53:48.490Z] Mismatch at 0: 30 != 129
```

**Review by junliume — DISMISSED** [review]

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -2069,18 +2069,33 @@ add_custom_test(smoke_solver_ConvHipImplicitGemmBwdDataV1R1 GFX103X_ENABLED TEST
 
 # MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_HIP_FWD_V4R1=1 is necessary due to WORKAROUND_iGemm_936 in Jenkinsfile,
 # which disables ConvHipImplicitGemmV4R1Fwd, but we still want to check that the solver is not broken.
-add_custom_test(smoke_solver_ConvHipImplicitGemmV4R1 GFX103X_ENABLED HALF_ENABLED BF16_ENABLED TEST_TUNING
+add_custom_test(smoke_solver_ConvHipImplicitGemmV4R1Fwd_fp32_fp16 GFX103X_ENABLED HALF_ENABLED TEST_TUNING
```
Comment (junliume) [inline test/CMakeLists.txt]:
```suggestion
add_custom_test(smoke_solver_ConvHipImplicitGemmV4R1Fwd_fp32 GFX103X_ENABLED TEST_TUNING
```

File: src/solver/conv_hip_implicit_gemm_fwd_v4r1.cpp [hunk 2]
```diff
@@ -61,10 +63,20 @@ bool ConvHipImplicitGemmV4R1Fwd::IsApplicable(const ConvolutionContext& ctx,
...
         return false;
-    if(ctx.GetStream().GetDeviceName() == "gfx90a" &&
-       problem.conv_problem.IsGfx90aFp16altRequired())
+    const auto device_name = ctx.GetStream().GetDeviceName();
+    if(device_name == "gfx90a" && problem.conv_problem.IsGfx90aFp16altRequired())
         return false;
 
+#if WORKAROUND_ISSUE_2038
+    if(!miopen::IsEnabled(MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_HIP_FWD_V4R1{}))
+        if(problem.IsBfp16())
```
Comment (junliume) [inline src/solver/conv_hip_implicit_gemm_fwd_v4r1.cpp]:
```suggestion
        if(problem.IsBfp16() || problem.IsFp16())
```

**Comment by atamazov** [comment]
@junliume @JehandadKhan @carlushuang @asroy PR is ready for the next round of reviewing and CI testing.

**Review by atamazov — COMMENTED** [review]
@junliume Thanks for 9ee84e088a87edb0645e82130959e1c39eea172a. It disables testing of HALF totally, but we still need to keep it for non-MI100/200 targets. Fixed now.

**Comment by atamazov** [comment]
Converted to draft due to https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2043#issuecomment-1483738387

**Comment by atamazov** [comment]
Seems not necessary, but let's keep this PR and branch for a month or so.

**Pull request closed without merging**